src/recommendations/src/recommendations-service/app.py

```diff
@@ -286,122 +320,217 @@ def recommendations():
+    """
+
+    app.logger.info(f"Items given for ranking: {items}")
+
+    # Extract item IDs from items supplied by caller. Note that unranked items
+    # can be specified as a list of objects with just an 'itemId' key or as a
+    # list of fully defined items/products (i.e. with an 'id' key).
+    item_map = {}
+    unranked_items = []
+    for item in items:
+        item_id = item.get('itemId') if item.get('itemId') else item.get('id')
+        item_map[item_id] = item
+        unranked_items.append(item_id)
+
+    app.logger.info(f"Unranked items: {unranked_items}")
+
+    resp_headers = {}
+    experiment = None
+    exp_manager = None
+
+    # Get active experiment if one is setup for feature.
+    if feature:
+        exp_manager = ExperimentManager()
+        experiment = exp_manager.get_active(feature)
+
+    if experiment:
+        app.logger.info('Using experiment: ' + experiment.name)
+
+        # Get ranked items from experiment.
+        tracker = exp_manager.default_tracker()
+
+        ranked_items = experiment.get_items(
+            user_id=user_id,
+            item_list=unranked_items,
+            tracker=tracker
+        )
+
+        app.logger.debug(f"Experiment ranking resolver gave us this ranking: {ranked_items}")
+
+        resp_headers['X-Experiment-Name'] = experiment.name
+        resp_headers['X-Experiment-Type'] = experiment.type
+        resp_headers['X-Experiment-Id'] = experiment.id
+    else:
+        # Fallback to default behavior of checking for campaign ARN parameter and
+        # then the default product resolver.
+        values = get_parameter_values([default_campaign_arn_param_name, filter_purchased_param_name])
+        app.logger.info(f'Falling back to Personalize: {values}')
+
+        campaign_arn = values[0]
+        filter_arn = values[1]
+
+        if campaign_arn:
+            resolver = PersonalizeRankingResolver(campaign_arn=campaign_arn, filter_arn=filter_arn)
+            resp_headers['X-Personalize-Recipe'] = get_recipe(campaign_arn)
+        else:
+            app.logger.info(f'Falling back to No-op: {values}')
+            resolver = RankingProductsNoOpResolver()
+
+        ranked_items = resolver.get_items(
+            user_id=user_id,
+            product_list=unranked_items
+        )
+
+    response_items = []
+    for ranked_item in ranked_items:
+
+        # Unlike with /recommendations and /related we are not hitting the products API to get product info back
+        # One would hope the caller would have it since the products are coming from there
+        item = item_map.get(ranked_item.get('itemId'))
+
+        if 'experiment' in ranked_item:
+
+            # We do not want to override an experiment if info already there
+            # e.g. if client is pulling recommendations then reranking them for example
+            # - this should not happen as the first recommendation pull would already be ranked
```

Comment:
The comment was trying to explain a potential corner case but you're absolutely right that it is unlikely anyone would try that - will delete the comment to avoid confusion.